### PR TITLE
InstancedMesh: more efficient initialization

### PR DESCRIPTION
**Pull request by ycw**
This PR uses the performant `copyWithin` to fill up `.instanceMatrix` on init. It has higher perf. than `setMatrixAt` approach [by this benchmark](https://stackblitz.com/edit/js-742cbf?file=index.js). The trick is that copies twice the amount for each iter:

| # | from | amount | to
|-|-|-|-
| 0 | 0..=15  | 16 | 16..=31
| 1 | 0..=31  | 32 | 32..=63
| 2 | 0..=63  | 64 | 64..=127
| 3 | 0..=127 | 128 | 128..=255
| .. | .. | .. | ..

:)

### CODE REVIEW

**Comment by Mugen87** [comment]
TBH, I've found the previous approach more readable. 

Did you encounter a performance problem when creating instances of `InstancedMesh`?

**Comment by ycw** [comment]
> readable

added .fill ... more readable now :)

> encounter a performance problem

slow first contentful paint

File: src/objects/InstancedMesh.js [hunk 1]
```diff
@@ -18,23 +18,15 @@ class InstancedMesh extends Mesh {
...
-		if ( array.length >= 16 ) array.set( _identity.elements );
-
-		for ( let i = 16, il = array.length; i < il; i <<= 1 ) {
-
-			array.copyWithin( i, 0, i );
-
-		}
-
-		this.instanceMatrix = new InstancedBufferAttribute( array, 16 );
+		this.instanceMatrix = new InstancedBufferAttribute( new Float32Array( count * 16 ), 16 );
```
Comment (mrdoob) [inline src/objects/InstancedMesh.js:21]: How about this?

```js
this.instanceMatrix = new InstancedBufferAttribute( new Float32Array( count * 16 ), 16 ).fill( _identity.elements );
```

Comment (ycw) [reply src/objects/InstancedMesh.js:21]: ok

File: src/core/BufferAttribute.js [hunk 2]
```diff
@@ -86,6 +86,27 @@ class BufferAttribute {
...
+		const end = Math.min( arrayLike.length, this.array.length );
+
+		for ( let i = 0; i < end; i ++ ) {
+
+			this.array[ i ] = arrayLike[ i ];
+
+		}
+
+		// copy twice the amount of previosu iteration
+		for ( let i = end, il = this.array.length; i < il; i <<= 1 ) {
```
Comment (Mugen87) [inline src/core/BufferAttribute.js:100]:
```
i <<= 1
```
TBH, I find this bitwise operation always hard to read. Can this be written in a different way?

Comment (ycw) [reply src/core/BufferAttribute.js:100]: ok.

File: src/core/BufferAttribute.js [hunk 3]
```diff
@@ -86,6 +86,26 @@ class BufferAttribute {
 
 	}
 
+	fill( arrayLike ) {
```
[the remark this answers is not in the document]
Comment (ycw) [reply src/core/BufferAttribute.js:89]: good.

**Review by Mugen87 — DISMISSED** [review]
Thanks for your work on this! The PR looks now good to me.

Although I favor simplicity in general, your benchmark shows a noticeable improvement so I think it's okay to accept the additional complexity.

**Comment by WestLangley** [comment]
>TBH, I've found the previous approach more readable.

I definitely agree.

>Although I favor simplicity in general

Me too. I'd also favor "clarity".

Is this being done only within the constructor? If so, this is not compelling IMHO. (I know this is a subjective decision.)

If it is every frame, that is a different matter.

**Comment by Mugen87** [comment]
> Is this being done only within the constructor? If so, this is not compelling IMHO. (I know this is a subjective decision.)

TBH, even after approving I did not feel 100% well with this PR. Maybe it's better to stall this change. I doubt this is a real performance bottleneck for apps.

**Comment by ycw** [comment]
oh, never mind.

**Comment by mrdoob** [comment]
I think part of the issue was that it was not very clear what were the benefits...

@ycw I do not know how to reat the table in the first post.

But in the benchmark I can see that initializing `InstancedMesh` with 1 million items goes from ~45ms to ~5ms, right?

**Comment by LeviPesin** [comment]
>  I do not know how to read the table in the first post.

I think it shows what is copied where in `copyWithin` calls.

**Pull request closed without merging**